### PR TITLE
Fix insert() 

### LINKED ISSUE #38: Change assessment of insert()
Homework Number W09H01

Citation of the homework instruction:
`insert()` Falls ein gleiches Tupel bereits vorhanden ist, wird das übergebene Tupel verworfen. Es soll das Tupel zurückgegeben werden, welches letztendlich im Pool gespeichert ist.

Additional Information:
If I understand `assertEquals()` correctly, it assesses the equality of the given tuples. For that, it uses our implementation of `tuple.equals(other)`. What we want though, is an assessment of identity.

This should fail, but does not:
`assertEquals(tuple0Duplicate, pool0.insert(tuple0Duplicate));`

I suggest, we change the test according to this example:
```java
assertEquals(tuple0, pool0.insert(tuple0Duplicate)); //old

/* assertEquals(expected, test of identity) */
assertEquals(true, pool0.insert(tuple0Duplicate) == tuple0);
assertEquals(false, pool0.insert(tuple0Duplicate) == tuple0Duplicate);
```

### PR DESCRIPTION
**Pull request by gonsoos**
Homework Number: W09H01

A short description of the changes proposed in the pull request:
    insert() should test for identity, not equality of tuples. #38 
    Fixed packages of classes, because Intellij was complaining (may be unnecessary).
    Added testcases for tuple, which has the attributes of an existing tuple, but in a reversed order.

[//]: <> (change [ ] to [x] in order to check the box)
- [x] Solution is not getting revealed
- [x] Files are formatted

### CODE REVIEW

File: W09H01/src/test/pgdp/hashset/TupleHashSetTest.java [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package pgdp.hashset;
+package test.pgdp.hashset;
```
Comment (LadnerJonas) [inline W09H01/src/test/pgdp/hashset/TupleHashSetTest.java]: Changing the package results in more work for earlier cloners. Everbody has to change the this line back, if he uses (sys)links (see README.md).

File: W09H01/src/test/pgdp/hashset/TupleHashSetTest.java [hunk 2]
```diff
@@ -9,6 +9,8 @@
 
 import static org.junit.jupiter.api.Assertions.*;
 
+import pgdp.hashset.TupleHashSet;
```
Comment (LadnerJonas) [inline W09H01/src/test/pgdp/hashset/TupleHashSetTest.java]: This is a result of the package change. Please revoke this change on every file

File: W09H01/src/test/pgdp/pools/TupleFactoryTest.java [hunk 3]
```diff
@@ -1,6 +1,7 @@
-package pgdp.pools;
+package test.pgdp.pools;
```
Comment (LadnerJonas) [inline W09H01/src/test/pgdp/pools/TupleFactoryTest.java]: Please undo this change.

File: W09H01/src/test/pgdp/pools/TuplePoolTest.java [hunk 4]
```diff
@@ -1,9 +1,9 @@
-package pgdp.pools;
+package test.pgdp.pools;
```
Comment (LadnerJonas) [inline W09H01/src/test/pgdp/pools/TuplePoolTest.java]: Please undo this change

File: W09H01/src/test/pgdp/pools/TuplePoolTest.java [hunk 5]
```diff
@@ -13,22 +13,25 @@ void insert() {
...
-    TuplePool<Integer, String> pool0 = new TuplePool<>();
-    
-    assertEquals(tuple0, pool0.insert(tuple0));
-    assertEquals(tuple1, pool0.insert(tuple1));
-    
-    assertEquals(tuple0, pool0.insert(tuple0Duplicate));
-    
-    assertEquals(tuple2, pool0.insert(tuple2));
+
+    Tuple<Object, Object> tuple0 = new Tuple<>(int0, string0);
```
Comment (LadnerJonas) [inline W09H01/src/test/pgdp/pools/TuplePoolTest.java:16]: Why do you exactly change the type to Object? Is there any benefit?

File: W09H01/src/test/pgdp/pools/TupleTest.java [hunk 6]
```diff
@@ -1,9 +1,9 @@
-package pgdp.pools;
+package test.pgdp.pools;
```
Comment (LadnerJonas) [inline W09H01/src/test/pgdp/pools/TupleTest.java]: Please undo this change

**Review by LadnerJonas — CHANGES_REQUESTED** [review]
- Changing the package results in more work for earlier cloners. Everbody has to change the this line back, if he uses (sys)links (see README.md).
- Please describe why changing some Tuples from Tuple<Integer, String> to Tuple<Object, Object> will benefit the testing.
- Everything else looks good 🏅

File: W09H01/src/test/pgdp/hashset/TupleHashSetTest.java (hunk 1 above)
Comment (gonsoos) [reply W09H01/src/test/pgdp/hashset/TupleHashSetTest.java]: Ok, I will!

File: W09H01/src/test/pgdp/pools/TuplePoolTest.java [hunk 7]
```diff
@@ -13,22 +13,25 @@ void insert() {
...
+    Tuple<Object, Object> tuple0Duplicate = new Tuple<>(int0, string0);
+    Tuple<Object, Object> tuple1 = new Tuple<>(int1, string1);
+
+    Tuple<Object, Object> tuple2 = new Tuple<>(int1, string0);
+    Tuple<Object, Object> tuple2Inverted = new Tuple<>(string0, int1);
+
+    TuplePool<Object, Object> pool0 = new TuplePool<>();
```
Comment (gonsoos) [inline W09H01/src/test/pgdp/pools/TuplePoolTest.java:23]: As I added a testcase for the insertion of `tuple2Inverted`, I needed to make sure that the `pool0` was able to handle both `Tuple<Integer, String>` and `Tuple<String, Integer>` objects. That's why I declared `pool0` as `TuplePool<Obj, Obj>`.
The UML diagram specifies that all methods and classes use the same generics `<T, S>`. Thus, if `TuplePool` uses `<Obj, Obj>`, all inserted tuples must be statically of type `Tuple<Object, Object>`. That's my understanding at least.

We could simply remove the `tuple2Inverted` testcase and revert all changes to the generics, though.

File: W09H01/src/test/pgdp/pools/TuplePoolTest.java (hunk 5 above)
Comment (gonsoos) [reply W09H01/src/test/pgdp/pools/TuplePoolTest.java:16]: See https://github.com/LadnerJonas/PGdP-Tests-WS21-22/pull/39/files/739a1b7efdd2be463ed934b4c92cabbd6968b110#r776915949

File: W09H01/src/test/pgdp/pools/TupleTest.java [hunk 8]
```diff
@@ -22,15 +22,17 @@ void testEquals() {
     Integer int0 = 123;
     String string0 = "Ladner";
     
-    Tuple<Integer, String> tuple0 = new Tuple<>(int0, string0);
+    Tuple<Object, Object> tuple0 = new Tuple<>(int0, string0);
     Tuple<Integer, String> tuple0Duplicate = new Tuple<>(int0, string0);
+    Tuple<Object, Object> tuple0Reversed = new Tuple<>(string0, int0);
```
Comment (gonsoos) [inline W09H01/src/test/pgdp/pools/TupleTest.java]: I'd revert the change of generics in `testEquals()`. Two tuples with different generics types should still be comparable.
(Apologies, if this is getting kind of convoluted. I'm still very unfamiliar with git(hub).)

**Pull request merged**